keras/callbacks.py

```diff
@@ -609,14 +611,16 @@ class TensorBoard(Callback):
         write_images: whether to write model weights to visualize as
             image in TensorBoard.
         embeddings_freq: frequency (in epochs) at which selected embedding
-            layers will be saved.
+            layers will be saved. If set to 0, embeddings won't be computed.
+            Data to be visualized must be passed as `embeddings_data`.
         embeddings_layer_names: a list of names of layers to keep eye on. If
             None or empty list all the embedding layer will be watched.
         embeddings_metadata: a dictionary which maps layer name to a file name
             in which metadata for this embedding layer is saved. See the
             [details](https://www.tensorflow.org/how_tos/embedding_viz/#metadata_optional)
             about metadata files format. In case if the same metadata file is
             used for all embedding layers, string can be passed.
+        embeddings_data: embedding data as Numpy array.
```

Comment:
Why is this argument necessary? The docstring doesn't describe what it is for or how to use it

Looking at the code, it seems that the data is used as model input to compute embeddings. Why not use the validation data?

On the technical side, this design also makes the assumption that the model has a single input, which maybe not be the case.

Reply:
Thanks for your feedback.

We need to define the `tf.Variable`s to hold the embeddings for every layer in `embeddings_layer_names` inside the `TensorBoard` callback's `set_model` method. The shape of these `embedding_var`s depends on 

1. the number of samples we want to embed (!) and
2. the embedding layer's output dimension

We don't have access to the `validation_data` in the `TensorBoard` callback's `set_model` method, though.

My first approach, up until commit 47f33a6bc411dd956dbfe8975df644d60e0d9666, was to pass an additional argument (`validation_data_size`) and use the `validation_data` to compute the embeddings. You asked to remove it (see above) because it's redundant and wouldn't be a good user experience to specify it. 

I agree.

@chleibig suggested to directly pass the data to be embedded as `embeddings_data` which would allow to embed any arbitrary data, be it

- the training data,
- the validation data,
- a subset thereof or
- something else entirely.

Now, passing `embeddings_data` is also kind of redundant because, in practice, it will probably just be the `validation_data`. I don't know which approach is better or worse but we need to know the number of samples to be embedded when we create the embedding variables in `set_model`. 

I'm happy to add more info about `embeddings_data` to the docstring if this is the way we go.

What do you think?

Reply:
Also, good catch about multiple inputs. I haven't really thought this through but it seems like kind of a special case right now..

**Do you have any specific example use cases mind?**

I could think of these scenarios:

- A siamese network (or similar)
  - multiple inputs of the same kind 
  - sharing the same network
- Shared embeddings
  - for example on images and words
  - multi-label classification using [label embedding](https://arxiv.org/pdf/1312.5650.pdf)
  - ... which I think requires a bit more work than simply taking the output of some arbitrary layer as the "embedding"
- Different inputs types
  - like the Twitter example of the Functional API Guide
    - sequence of words + time of the day
    - ... not sure what to actually visualize in this case
  - or the visual question answering model
- [...]

I think this requires some more thought.

What do you think about tackling this in a separate Issue/PR?

Reply:
Hi together and sorry @dschwertfeger that my suggestion made you having to argue about it;)

I completely agree that the least "redundant" version should be favourable.

For completeness, I list here another reason why I had suggested `embeddings_data`: in terms of style it would be in line with the keyword `embeddings_metadata`,  which provides the labels or other information such as thumbnail images. Furthermore, in case of large (good for evaluation purposes) validation data sets, it may be impractical - due to the memory consumption - to embed all of it. Being able to pass only a subset of the validation data via `embeddings_data` may hence be more desirable.

In case of multiple inputs, `embeddings_metadata` should pass all of them I guess, such as documented e.g. [here](https://keras.io/getting-started/functional-api-guide/).